dll/win32/kernel32/kernel32_vista/firmware.c

```diff
@@ -0,0 +1,231 @@
+/*
+ * PROJECT:     ReactOS Win32 Base API
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     NT6+ Firmware API
+ * COPYRIGHT:   Copyright 2024 Ratin Gao <ratin@knsoft.org>
+ */
+
+/* We need NT6+ Ex definitions */
+#undef NTDDI_VERSION
+#define NTDDI_VERSION NTDDI_VISTA
+#include <ndk/exfuncs.h>
+
+#include "k32_vista.h"
+
+_Success_(return > 0)
+DWORD
+WINAPI
+GetFirmwareEnvironmentVariableExW(
+    _In_ LPCWSTR lpName,
+    _In_ LPCWSTR lpGuid,
+    _Out_writes_bytes_to_opt_(nSize, return) PVOID pBuffer,
+    _In_ DWORD nSize,
+    _Out_opt_ PDWORD pdwAttribubutes)
+{
+    NTSTATUS Status;
+    UNICODE_STRING VariableName, Namespace;
+    GUID VendorGuid;
+    ULONG Length;
+
+    /* Check input parameters and build NT strings */
+    if (lpName == NULL || lpGuid == NULL)
```

Comment:
```suggestion
    if (!lpName || !lpGuid)
```

Reply:
Imo, use `!` to compare `BOOL`, `NULL` to compare pointer, `0` to compare number is more make sense.

Reply:
I am here with @HBelusca . As long as the variables name indicates via its hungarian that it is a pointer, then checking for NULL like he suggested, gives very compact, and still correct and readable code.
I can understand your concerns @RatinCN (but **just** for cases where no hungarian is used)

Reply:
Committed.